src/wast-lexer.cc

```diff
@@ -32,27 +32,77 @@
 
-#define YY_USER_ACTION                  \
-  {                                     \
-    loc->filename = filename_;          \
-    loc->line = line_;                  \
-    loc->first_column = COLUMN(token_); \
-    loc->last_column = COLUMN(cursor_); \
-  }
+#define PUSH_LOOKAHEAD_TOKEN              \
+  do {                                    \
+    WastLexer::LexToken tok;              \
+    lookahead_->tokens_.push_back(tok);   \
+    token_ = &lookahead_->tokens_.back(); \
+  } while (0)
+
+#define YY_USER_ACTION(loc)                \
+  (loc)->filename = filename_;             \
+  (loc)->line = line_;                     \
+  (loc)->first_column = COLUMN(next_pos_); \
+  (loc)->last_column = COLUMN(cursor_);
+
+#define NAME_TO_VALUE(name) WABT_TOKEN_TYPE_##name
+
+#define TOKEN_VALUE(value)         \
+    YY_USER_ACTION(&token_->loc_); \
+    token_->value_ = value;        \
+    next_pos_ = cursor_;
+
+#define LOOKAHEAD_VALUE(value) \
+    TOKEN_VALUE(value);        \
+    PUSH_LOOKAHEAD_TOKEN;
+
+#define LOOKAHEAD(name)            \
+  int value = NAME_TO_VALUE(name); \
+  LOOKAHEAD_VALUE(value);          \
+  next_pos_ = cursor_;             \
+  continue
+
+#define RETURN_TOKEN                                         \
+  do {                                                       \
+    WastLexer::LexToken* tok = &lookahead_->tokens_.front(); \
+    *loc = tok->loc_;                                        \
+    *lval = tok->lval_;                                      \
+    int Result = tok->value_;                                \
+    lookahead_->tokens_.pop_front();                         \
+    if (lookahead_->tokens_.empty()) token_ = nullptr;       \
+    return Result;                                           \
+  } while (0)
 
-#define RETURN(name) \
-  YY_USER_ACTION;    \
-  return WABT_TOKEN_TYPE_##name
+#define RETURN(name)                 \
+  do {                               \
+    int value = NAME_TO_VALUE(name); \
+    TOKEN_VALUE(value);              \
+    RETURN_TOKEN;                    \
+  } while (0)
+
+#define RETURN_LPAR(name) \
+  do { \
+    int value = NAME_TO_VALUE(name);      \
+    TOKEN_VALUE(value);                   \
+    if (lookahead_is_lpar()) {            \
+      value = NAME_TO_VALUE(LPAR_##name); \
+      YY_USER_ACTION(loc);                \
+      return value;                       \
+    }                                     \
+    RETURN_TOKEN;                         \
+  } while (0)
 
-#define ERROR(...) \
-  YY_USER_ACTION;  \
+#define ERROR(...)                                  \
+  YY_USER_ACTION(loc);                              \
   wast_parser_error(loc, this, parser, __VA_ARGS__)
 
 #define BEGIN(c) cond = (c)
 #define FILL(n)                                \
   do {                                         \
     if (WABT_FAILED(Fill(loc, parser, (n)))) { \
-      RETURN(EOF);                             \
+      int value = NAME_TO_VALUE(EOF);          \
```

Comment:
Unfortunately, a macro is "tokenized" first, and "#" concatenates tokens. Hence, when nested, it converts the token sequence to "RETURN "(EOF" ")". It took me a long time to realize that this was the main reason I was having trouble with many of the previous macros